kubernetes/schema_pod_spec.go

```diff
@@ -33,7 +33,38 @@ func podSpecFields(isUpdatable bool) map[string]*schema.Schema {
 			Type:        schema.TypeString,
 			Optional:    true,
 			Default:     "ClusterFirst",
-			Description: "Set DNS policy for containers within the pod. One of 'ClusterFirst' or 'Default'. Defaults to 'ClusterFirst'.",
+			Description: "Set DNS policy for containers within the pod. One of 'ClusterFirst', 'Default', 'ClusterFirstWithHostNet', or 'None'. Defaults to 'ClusterFirst'.",
+		},
+		"dns_config": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			MaxItems:    1,
+			Description: "DnsConfig holds pod-level DNS settings. Optional: Defaults to empty",
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"nameservers": {
+						Type:        schema.TypeSet,
+						Description: "A list of IP addresses that will be used as DNS servers for the Pod. There can be at most 3 IP addresses specified. When the Pod’s dnsPolicy is set to “None”, the list must contain at least one IP address, otherwise this property is optional. The servers listed will be combined to the base nameservers generated from the specified DNS policy with duplicate addresses removed.",
+						Optional:    true,
+						MaxItems:    3,
+						Elem:        &schema.Schema{Type: schema.TypeString},
+						Set:         schema.HashString,
+					},
+					"searches": {
+						Type:        schema.TypeSet,
+						Description: "A list of DNS search domains for hostname lookup in the Pod. This property is optional. When specified, the provided list will be merged into the base search domain names generated from the chosen DNS policy. Duplicate domain names are removed. Kubernetes allows for at most 6 search domains.",
+						Optional:    true,
+						MaxItems:    6,
+						Elem:        &schema.Schema{Type: schema.TypeString},
+						Set:         schema.HashString,
+					},
+					"options": {
```

Comment:
This should actually be a list of "name" & "value" pairs. Defining it as a `schema.TypeMap` allows for any "foo" & "bar" pair to be a valid element and Terraform won't be able to detect the wrong key names through it's schema validation.

This means, the `options` element itself should be of type `schema.TypeList` and it's `Elem` property again be a resource of type `schema.TypeList` with members `name` and `value` (much like `dns_config` itself is configured). You can then enforce `name` to be mandatory and value to be optional. You would however need to detect duplicates in the list yourself, in the validation function (don't remove them, just error out).

A second option (which I haven't actually tried out) is to make `options` be `schema.TypeSet` and its `Elem`-ents be `schema.TypeList` with `name` and `value` attributes configured like above. You would however need to set the `Set` attribute on `options` to `schema.HashResource` to make sure the contents of its `Elem`s are properly accounted for as a set.

I hope this makes some amount of sense.
I'm here to help out further. Just let me know.